astropy/stats/spatial.py

```diff
@@ -0,0 +1,314 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+This module implements functions and classes for spatial statistics.
+"""
+
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+import numpy as np
+import math
+
+class RipleysKEstimator(object):
+    """
+    Estimators for Ripley's K function for two-dimensional spatial data.
+
+    Parameters
+    ----------
+    area : float
+        Area of study from which the points where observed.
+    x_max, y_max : float, float, optional
+        Maximum rectangular coordinates of the area of study.
+        Required if ``mode == 'translation'`` or ``mode == ohser``.
+    x_min, y_min : float, float, optional
+        Minimum rectangular coordinates of the area of study.
+        Required if ``mode == 'variable-width'`` or ``mode == ohser``.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from matplotlib import pyplot as plt # doctest: +SKIP
+    >>> from astropy.stats import RipleysKEstimator
+    >>> z = np.random.uniform(low=5, high=10, size=(100, 2))
+    >>> Kest = RipleysKEstimator(area=25, x_max=10, y_max=10,
+    ... x_min=5, y_min=5)
+    >>> r = np.linspace(0, 2.5, 100)
+    >>> plt.plot(r, Kest.poisson(r)) # doctest: +SKIP
+    >>> plt.plot(r, Kest(data=z, radii=r, mode='none')) # doctest: +SKIP
+    >>> plt.plot(r, Kest(data=z, radii=r, mode='translation')) # doctest: +SKIP
+    >>> plt.plot(r, Kest(data=z, radii=r, mode='ohser')) # doctest: +SKIP
+    >>> plt.plot(r, Kest(data=z, radii=r, mode='var-width')) # doctest: +SKIP
+    >>> plt.plot(r, Kest(data=z, radii=r, mode='ripley')) # doctest: +SKIP
+
+    References
+    ----------
+    .. [1] Spatial descriptive statistics.
+       <https://en.wikipedia.org/wiki/Spatial_descriptive_statistics>
+    .. [2] Package spatstat.
+       <https://cran.r-project.org/web/packages/spatstat/spatstat.pdf>
+    .. [3] Cressie, N.A.C. (1991). Statistics for Spatial Data,
+       Wiley, New York.
+    .. [4] Stoyan, D., Stoyan, H. (1992). Fractals, Random Shapes and
+       Point Fields, Akademie Verlag GmbH, Chichester.
+    """
+
+    def __init__(self, area, x_max=None, y_max=None, x_min=0, y_min=0):
+        self.area = area
+        self.x_max = x_max
+        self.y_max = y_max
+        self.x_min = x_min
+        self.y_min = y_min
+
+    @property
+    def area(self):
+        return self._area
+
+    @area.setter
+    def area(self, value):
+        if isinstance(value, (float, int)) and value > 0:
+            self._area = value
+        else:
+            raise ValueError('area is expected to be a positive number. '
+                             'Got {}.'.format(value))
+
+    @property
+    def y_max(self):
+        return self._y_max
+
+    @y_max.setter
+    def y_max(self, value):
+        if value is None or (isinstance(value, (float, int)) and value > 0):
```

Comment:
That's mostly a stylistic comment, but I generally prefer `if ... elif ... else` in such cases:

```
if value is None:
    self._y_max = None
elif isinstance(value, (float, int)) and value > 0:
    self._y_max = value
else:
    # err
```